Source/ACE/Network/NetworkSession.cs

```diff
@@ -12,110 +13,124 @@ namespace ACE.Network
         {
             this.session = session;
-            ConnectionData  = new SessionConnectionData();
-            currentBundle = new NetworkBundle();
-            nextSend = DateTime.UtcNow;
-            nextResync = DateTime.UtcNow;
-            nextAck = DateTime.UtcNow;
         }
 
         public void EnqueueSend(params GameMessage[] messages)
         {
-            currentBundle.EncryptedChecksum = true;
-            foreach (var message in messages)
-                currentBundle.Enqueue(message);
+            lock (lockObject)
+            {
+                currentBundle.EncryptedChecksum = true;
+
+                foreach (var message in messages)
+                    currentBundle.Enqueue(message);
+            }
         }
 
         public void EnqueueSend(params ServerPacket[] packets)
         {
-            foreach (var packet in packets)
-                PacketQueue.Enqueue(packet);
+            lock (lockObject)
+            {
+                foreach (var packet in packets)
+                    packetQueue.Enqueue(packet);
+            }
         }
 
         public void Update(double lastTick)
         {
-            ConnectionData.ServerTime += lastTick;
-
-            if (sendResync && !currentBundle.TimeSync && DateTime.UtcNow > nextResync)
+            lock (lockObject)
             {
-                currentBundle.TimeSync = true;
-                currentBundle.EncryptedChecksum = true;
-                nextResync = DateTime.UtcNow.AddMilliseconds(timeBetweenTimeSync);
-            }
+                ConnectionData.ServerTime += lastTick;
 
-            if (sendAck && !currentBundle.SendAck && DateTime.UtcNow > nextAck)
-            {
-                currentBundle.SendAck = true;
-                nextAck = DateTime.UtcNow.AddMilliseconds(timeBetweenAck);
-            }
+                if (sendResync && !currentBundle.TimeSync && DateTime.UtcNow > nextResync)
+                {
+                    currentBundle.TimeSync = true;
+                    currentBundle.EncryptedChecksum = true;
+                    nextResync = DateTime.UtcNow.AddMilliseconds(timeBetweenTimeSync);
+                }
 
-            if (currentBundle.NeedsSending && DateTime.UtcNow > nextSend)
-            {
-                FlushBundle();
-            }
+                if (sendAck && !currentBundle.SendAck && DateTime.UtcNow > nextAck)
+                {
+                    currentBundle.SendAck = true;
+                    nextAck = DateTime.UtcNow.AddMilliseconds(timeBetweenAck);
+                }
+
+                if (currentBundle.NeedsSending && DateTime.UtcNow > nextSend)
+                {
+                    FlushBundle();
+                }
 
-            FlushPackets();
+                FlushPackets();
+            }
         }
 
         public void HandlePacket(ClientPacket packet)
         {
-            lastReceivedSequence = packet.Header.Sequence;
+            lock (lockObject)

```

Comment:
I really don't see the need for this lock.  If stuff has to Enqueue as a response, it will hit those locks accordingly.  This lock could significantly degrade performance.